File: internal/model/scheduler.go
package model

import (
	"time"
)

// TaskQueue 任务队列
type TaskQueue struct {
	ID          uint   `json:"id" gorm:"primaryKey"`
	Name        string `json:"name" gorm:"uniqueIndex;size:64;not null"`
	DisplayName string `json:"display_name" gorm:"size:128"`
	Description string `json:"description" gorm:"size:512"`

	// 队列配置
	Priority int `json:"priority" gorm:"default:0"`     // 队列优先级
	Weight   int `json:"weight" gorm:"default:1"`       // 资源分配权重

	// 资源配额
	MaxCPU    string `json:"max_cpu" gorm:"size:32"`    // 最大 CPU
	MaxMemory string `json:"max_memory" gorm:"size:32"` // 最大内存
	MaxGPU    int    `json:"max_gpu"`                    // 最大 GPU 数量
	MaxTasks  int    `json:"max_tasks" gorm:"default:100"` // 最大任务数

	// 调度策略
	Policy     string `json:"policy" gorm:"size:32;default:'fifo'"` // fifo, priority, fair
	Preemption bool   `json:"preemption" gorm:"default:false"`      // 是否允许抢占

	// 状态
	Status    string `json:"status" gorm:"size:20;default:'active'"` // active, paused, disabled
	TaskCount int    `json:"task_count" gorm:"-"`                    // 运行中任务数

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

func (TaskQueue) TableName() string {
	return "task_queues"
}

// Task 任务定义
type Task struct {
	ID     uint   `json:"id" gorm:"primaryKey"`
	TaskID string `json:"task_id" gorm:"column:task_uid;uniqueIndex;size:64;not null"` // 任务唯一标识
	Name   string `json:"name" gorm:"size:128;not null"`
	// 任务归属
	UserID    uint `json:"user_id" gorm:"index;not null"`
	User      User `json:"user" gorm:"foreignKey:UserID"`
	QueueID   uint `json:"queue_id" gorm:"index;not null"`
	Queue     TaskQueue `json:"queue" gorm:"foreignKey:QueueID"`
	ClusterID uint `json:"cluster_id" gorm:"index;not null"`
	Cluster   Cluster `json:"cluster" gorm:"foreignKey:ClusterID"`

	// 任务类型
	TaskType string `json:"task_type" gorm:"size:32;not null"` // job, cronjob, volcano, mpi, pytorch
	Priority int    `json:"priority" gorm:"default:0"`         // 任务优先级 0-1000

	// 资源需求
	CPU     string `json:"cpu" gorm:"size:32"`
	Memory  string `json:"memory" gorm:"size:32"`
	GPU     int    `json:"gpu"`
	GPUType string `json:"gpu_type" gorm:"size:32"`

	// Gang Scheduling 配置
	MinReplicas int `json:"min_replicas" gorm:"default:1"`
	Replicas    int `json:"replicas" gorm:"default:1"`

	// 任务配置
	Image      string `json:"image" gorm:"size:256"`
	Command    string `json:"command" gorm:"type:text"`
	Args       string `json:"args" gorm:"type:text"`
	EnvVars    string `json:"env_vars" gorm:"type:text"`
	VolumeMounts string `json:"volume_mounts" gorm:"type:text"`

	// 调度约束
	NodeSelector string `json:"node_selector" gorm:"type:text"`
	Tolerations  string `json:"tolerations" gorm:"type:text"`

	// 超时与重试
	Timeout    int `json:"timeout" gorm:"default:3600"`
	MaxRetry   int `json:"max_retry" gorm:"default:3"`
	RetryCount int `json:"retry_count"`

	// 状态
	Status  string `json:"status" gorm:"size:32;default:'pending'"` // pending, queued, running, succeeded, failed, cancelled
	Message string `json:"message" gorm:"type:text"`

	// K8S 资源引用
	Namespace  string `json:"namespace" gorm:"size:64;default:'default'"`
	K8SJobName string `json:"k8s_job_name" gorm:"size:128"`

	// 时间戳
	SubmittedAt *time.Time `json:"submitted_at"`
	StartedAt   *time.Time `json:"started_at"`
	CompletedAt *time.Time `json:"completed_at"`
	CreatedAt   time.Time  `json:"created_at"`
	UpdatedAt   time.Time  `json:"updated_at"`
}

func (Task) TableName() string {
	return "tasks"
}

// TaskLog 任务日志
type TaskLog struct {
	ID        uint      `json:"id" gorm:"primaryKey"`
	TaskID    uint      `json:"task_id" gorm:"column:task_ref_id;index"`
	Level     string    `json:"level" gorm:"size:20"` // info, warn, error
	Message   string    `json:"message" gorm:"type:text"`
	CreatedAt time.Time `json:"created_at"`
}

func (TaskLog) TableName() string {
	return "task_logs"
}

// ResourceReservation 资源预留
type ResourceReservation struct {
	ID        uint   `json:"id" gorm:"primaryKey"`
	Name      string `json:"name" gorm:"size:128;not null"`
	UserID    uint   `json:"user_id" gorm:"index;not null"`
	User      User   `json:"user" gorm:"foreignKey:UserID"`
	QueueID   uint   `json:"queue_id" gorm:"index;not null"`
	Queue     TaskQueue `json:"queue" gorm:"foreignKey:QueueID"`
	ClusterID uint   `json:"cluster_id" gorm:"index;not null"`
	Cluster   Cluster `json:"cluster" gorm:"foreignKey:ClusterID"`

	// 预留资源
	CPU     string `json:"cpu" gorm:"size:32"`
	Memory  string `json:"memory" gorm:"size:32"`
	GPU     int    `json:"gpu"`
	GPUType string `json:"gpu_type" gorm:"size:32"`

	// 预留策略
	StartTime time.Time `json:"start_time"`
	EndTime   time.Time `json:"end_time"`
	Recurring bool      `json:"recurring"`
	CronExpr  string    `json:"cron_expr" gorm:"size:64"`

	// 节点绑定
	NodeName     string `json:"node_name" gorm:"size:128"`
	NodeSelector string `json:"node_selector" gorm:"type:text"`

	// 状态
	Status string `json:"status" gorm:"size:20;default:'active'"` // active, expired, cancelled

	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

func (ResourceReservation) TableName() string {
	return "resource_reservations"
}

// SchedulePolicy 调度策略
type SchedulePolicy struct {
	ID          uint   `json:"id" gorm:"primaryKey"`
	Name        string `json:"name" gorm:"uniqueIndex;size:64;not null"`
	Description string `json:"description" gorm:"size:256"`

	// 策略类型
	Type   string `json:"type" gorm:"size:32;not null"` // priority, fair, gang, binpack, spread
	Config string `json:"config" gorm:"type:text"`       // JSON 配置

	// 适用范围
	ClusterID *uint `json:"cluster_id"`
	QueueID   *uint `json:"queue_id"`

	IsDefault bool      `json:"is_default" gorm:"default:false"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

func (SchedulePolicy) TableName() string {
	return "schedule_policies"
}

File: frontend/src/api/scheduler.ts
import { get, post, put, del } from './request'

// ==================== 队列管理 ====================

export interface TaskQueue {
  id: number
  name: string
  display_name: string
  description: string
  priority: number
  weight: number
  max_cpu: string
  max_memory: string
  max_gpu: number
  max_tasks: number
  policy: string
  preemption: boolean
  status: string
  running_tasks: number
  pending_tasks: number
  created_at: string
}

export interface CreateQueueRequest {
  name: string
  display_name?: string
  description?: string
  priority?: number
  weight?: number
  max_cpu?: string
  max_memory?: string
  max_gpu?: number
  max_tasks?: number
  policy?: string
  preemption?: boolean
}

export const listQueues = () => {
  return get<{ code: number; data: TaskQueue[] }>('/scheduler/queues')
}

export const createQueue = (data: CreateQueueRequest) => {
  return post<{ code: number; data: TaskQueue }>('/scheduler/queues', data)
}

export const getQueue = (id: number) => {
  return get<{ code: number; data: { queue: TaskQueue; running_tasks: number; pending_tasks: number; total_tasks: number } }>(`/scheduler/queues/${id}`)
}

export const updateQueue = (id: number, data: Partial<CreateQueueRequest>) => {
  return put(`/scheduler/queues/${id}`, data)
}

export const deleteQueue = (id: number) => {
  return del(`/scheduler/queues/${id}`)
}

// ==================== 任务管理 ====================

export interface Task {
  id: number
  task_id: string
  name: string
  user_id: number
  queue_id: number
  cluster_id: number
  task_type: string
  priority: number
  cpu: string
  memory: string
  gpu: number
  gpu_type: string
  min_replicas: number
  replicas: number
  image: string
  command: string
  args: string
  env_vars: string
  namespace: string
  timeout: number
  max_retry: number
  retry_count: number
  status: string
  message: string
  k8s_job_name: string
  submitted_at: string
  started_at: string
  completed_at: string
  created_at: string
  queue?: TaskQueue
  user?: { id: number; username: string; real_name: string }
}

export interface TaskLog {
  id: number
  task_id: number
  level: string
  message: string
  created_at: string
}

export interface CreateTaskRequest {
  name: string
  queue_id: number
  cluster_id: number
  task_type: string
  priority?: number
  cpu?: string
  memory?: string
  gpu?: number
  gpu_type?: string
  replicas?: number
  min_replicas?: number
  image: string
  command?: string[]
  args?: string[]
  env_vars?: Record<string, string>
  namespace?: string
  timeout?: number
  max_retry?: number
}

export const listTasks = (params?: {
  page?: number
  size?: number
  queue_id?: number
  status?: string
}) => {
  return get<{ code: number; data: Task[]; total: number; page: number; size: number }>('/scheduler/tasks', { params })
}

export const createTask = (data: CreateTaskRequest) => {
  return post<{ code: number; data: Task }>('/scheduler/tasks', data)
}

export const getTask = (id: number) => {
  return get<{ code: number; data: { task: Task; logs: TaskLog[] } }>(`/scheduler/tasks/${id}`)
}

export const cancelTask = (id: number) => {
  return post(`/scheduler/tasks/${id}/cancel`)
}

export const retryTask = (id: number) => {
  return post(`/scheduler/tasks/${id}/retry`)
}

export const getTaskLogs = (id: number) => {
  return get<{ code: number; data: TaskLog[] }>(`/scheduler/tasks/${id}/logs`)
}

// ==================== 资源预留 ====================

export interface ResourceReservation {
  id: number
  name: string
  user_id: number
  queue_id: number
  cluster_id: number
  cpu: string
  memory: string
  gpu: number
  gpu_type: string
  start_time: string
  end_time: string
  recurring: boolean
  cron_expr: string
  node_name: string
  node_selector: string
  status: string
  created_at: string
}

export interface CreateReservationRequest {
  name: string
  queue_id: number
  cluster_id: number
  cpu?: string
  memory?: string
  gpu?: number
  gpu_type?: string
  start_time: string
  end_time: string
  recurring?: boolean
  cron_expr?: string
  node_name?: string
  node_selector?: Record<string, string>
}

export const listReservations = (params?: {
  user_id?: number
  status?: string
}) => {
  return get<{ code: number; data: ResourceReservation[] }>('/scheduler/reservations', { params })
}

export const createReservation = (data: CreateReservationRequest) => {
  return post<{ code: number; data: ResourceReservation }>('/scheduler/reservations', data)
}

export const deleteReservation = (id: number) => {
  return del(`/scheduler/reservations/${id}`)
}
